go-controller/pkg/node/controllers/egressip/egressip.go

```diff
@@ -0,0 +1,928 @@
+			c.eIPQueue.AddRateLimited(key)
+			return true
+		}
+		klog.Errorf("Dropping Egress IP %q out of the queue: %w", key, err)
+		utilruntime.HandleError(err)
+	}
+	c.eIPQueue.Forget(key)
+	return true
+}
+
+func (c *Controller) syncEIP(eIPName string) error {
+	// 1. Lock on the existing 'state', as we are going to use it for cleanup and update.
+	// 2. Build latest 'config'. This includes listing referenced namespaces and pods.
+	// To make sure there is no race with pod and namespace handlers, referencedObjects is acquired
+	// before listing objects, and released when the 'config' is built. At this point namespace and pod
+	// handler can use referencedObjects to see which objects were considered as related by the handler last time.
+	// 3. With existing state and newly generated config, we can clean up and apply.
+	return c.cache.DoWithLock(eIPName, func(eIPName string) error {
+		informerEIP, err := c.eIPLister.Get(eIPName)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get Egress IP before sync: %w", err)
+		}
+		var update *config
+		// get updated policy and update policy refs
+		if apierrors.IsNotFound(err) || (informerEIP != nil && !informerEIP.DeletionTimestamp.IsZero()) {
+			// EIP deleted
+			update = nil
+			c.deleteRefObjects(eIPName)
+		} else {
+			update, err = c.getConfigAndUpdateRefs(informerEIP, true)
+			if err != nil {
+				return fmt.Errorf("failed to get config and update references for Egress IP %s: %w", eIPName, err)
+			}
+		}
+		existing, found := c.cache.Load(eIPName)
+		if !found {
+			if update == nil {
+				// nothing to do
+				return nil
+			}
+			existing = newState()
+			c.cache.Store(eIPName, existing)
+		}
+		if err = c.updateEIP(existing, update); err != nil {
+			return fmt.Errorf("failed to update policy from %+v to %+v: %w", existing, update, err)
+		}
+		if update == nil {
+			c.cache.Delete(eIPName)
+		}
+		return nil
+	})
+}
+
+// getConfigAndUpdateRefs lists and updates all referenced objects for a given EIP and returns
+// config to perform an update.
+// This function should be the only one that lists referenced objects, and updates referencedObjects atomically.
+func (c *Controller) getConfigAndUpdateRefs(eIP *eipv1.EgressIP, updateRefs bool) (*config, error) {
+	c.referencedObjectsLock.Lock()
+	defer c.referencedObjectsLock.Unlock()
+	eIPConfig, selectedNamespaces, selectedPods, err := c.processEIP(eIP)
+	if err != nil {
+		return nil, err
+	}
+	// attempt to detect overlapping EIPs - reject EIP if there's overlap
+	targetNs, err := c.listNamespacesBySelector(&eIP.Spec.NamespaceSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list target namespaces: %w", err)
+	}
+	targetNsNames := sets.Set[string]{}
+	targetNamespaces := map[string]map[ktypes.NamespacedName]*corev1.Pod{}
+	for _, ns := range targetNs {
+		targetNsNames.Insert(ns.Name)
+		targetPods, err := c.podLister.Pods(ns.Name).List(labels.Everything())
```

Comment:
I don't understand the disparity of this `targetPods` here vs `selectedPods` above. Should it be the same? If not, perhaps a comment would help explaining why.

Reply:
it was wrong. I removed it.